File: Chapter11/ThePeopleWhoChat.Core/ChatServiceClient.fs
namespace ThePeopleWhoChat.Core

    open System
    open System.Text
    open System.Net
    open System.IO
    open System.Web
    open System.Runtime.Serialization.Json

    type ServiceClient(url:string) =

        member private x.JsonSerialize<'T>(data:'T) =
            use ms = new MemoryStream()
            let ser = new DataContractJsonSerializer(typeof<'T>)
            ser.WriteObject(ms, data)
            ms.Seek(0L,SeekOrigin.Begin) |> ignore
            use sr = new StreamReader(ms)
            sr.ReadToEnd()

        member private x.makeRequest<'T>(token:string,verb:string,url:string,data:'T) =
            let req = WebRequest.Create(new Uri(url)) :?> HttpWebRequest
            req.Method <- verb
            req.Headers.Add(Consts.TokenHeaderName,token)      
            let json = x.JsonSerialize(data)      
            if not (typeof<'T> = typeof<unit>) then
                let buffer = Encoding.UTF8.GetBytes(json)                
                req.ContentType <- "application/json"
                req.ContentLength <- buffer.LongLength
                use reqSt = req.GetRequestStream()
                reqSt.Write(buffer,0,buffer.Length)
            req
        member private x.WebRequestWrapper<'U>(f:unit->'U) =
            try
                f()
            with
            | :? WebException as e ->
                    let resp = e.Response :?> HttpWebResponse
                    let err = resp.Headers.[Consts.ErrorHeaderName]
                    let msg = match box err with
                                | null -> "Service failure"
                                | _ -> err
                    failwith (sprintf "%s: status = %d (%s)" msg (int32 resp.StatusCode) (resp.StatusCode.ToString()))
        member private x.getResponse<'U>(req:HttpWebRequest) =
            x.WebRequestWrapper(fun () ->
                let res = req.GetResponse() :?> HttpWebResponse
                use resSt = res.GetResponseStream()
                let ser = DataContractJsonSerializer(typeof<'U>)
                ser.ReadObject(resSt) :?> 'U
            )
        member private x.getResponseUnit(req:HttpWebRequest) =
            x.WebRequestWrapper(fun () ->
                let res = req.GetResponse() :?> HttpWebResponse
                ())
        member private x.getData<'U>(token:string,path:string) =
            let fullPath = String.Format("{0}/{1}",url,path)
            let req = x.makeRequest(token,"GET",fullPath,())
            x.getResponse<'U>(req)
        member private x.putDataUnit<'T>(token:string,path:string,data:'T) =
            let fullPath = String.Format("{0}/{1}",url,path)
            let req = x.makeRequest(token,"PUT",fullPath,data)
            x.getResponseUnit(req)
        member private x.putData<'T,'U>(token:string,path:string,data:'T) =
            let fullPath = String.Format("{0}/{1}",url,path)
            let req = x.makeRequest(token,"PUT",fullPath,data)
            x.getResponse<'U>(req)
        member private x.postData<'T,'U>(token:string,path:string,data:'T) =
            let fullPath = String.Format("{0}/{1}",url,path)
            let req = x.makeRequest(token,"POST",fullPath,data)
            x.getResponse<'U>(req)
        member private x.deleteData(token:string,path:string) =
            let fullPath = String.Format("{0}/{1}",url,path)
            let req = x.makeRequest(token,"DELETE",fullPath,())
            x.getResponseUnit(req)

        interface IChatServiceClient with

            member this.Login(username:string,password:string) =
                let logon = {username = username; password = password}
                let req = this.makeRequest("","PUT",String.Format("{0}/token",url),logon)
                let res = this.WebRequestWrapper(fun () ->
                    req.GetResponse() :?> HttpWebResponse
                    )
                res.Headers.[Consts.TokenHeaderName]

            member this.Logout(token:string) =
                this.deleteData(token,String.Format("token/{0}",token))

            member this.ListSessions(token:string) =
                this.getData<LoginSession array>(token,"sessions")

            member this.AddUser(token:string, user:User) =
                this.putData<User,string>(token,"users",user)

            member this.RemoveUser(token:string, userId:string) =
                if userId.StartsWith("users/") then this.deleteData(token,userId)
                else failwith (sprintf "invalid userId: %s" userId)

            member this.ListUsers(token:string) =
                this.getData<User array>(token,"users")

            member this.AddRoom(token:string, room:Room) =
                this.putData<Room,string>(token,"rooms",room)

            member this.RemoveRoom(token:string, roomId:string) =
                if roomId.StartsWith("rooms/") then this.deleteData(token,roomId)
                else failwith (sprintf "invalid roomId: %s" roomId)

            member this.ListRooms(token:string) =
                this.getData<Room array>(token, "rooms")
                
            member this.EnterRoom(token:string, roomId:string) =
                this.putDataUnit<Identifier>(token,"currentroom",{id = roomId})

            member this.LeaveRoom(token:string) =
                this.deleteData(token,"currentroom")

            member this.GetMessages(token:string, from:DateTime) =
                this.getData<Message array>(token, String.Format("messages?after={0:yyyyMMddHHmmssfffffff}",from))

            member this.PostMessage(token:string, message:string) =
                this.putDataUnit<MessageText>(token,"messages",{message = message})

File: Chapter11/ThePeopleWhoChat.Web/Scripts/ChatServiceClient.js

var baseUrl = "/Chat.svc/";

function service_redirectIfNotLoggedIn(redirect) {
    if (sessionStorage.token == null) { redirect() }
}

function service_login(username, password, success, failure) {
    var login = { username: username, password: password };
    $.ajax({
        type: "PUT",
        url: baseUrl +"token",
        data: JSON.stringify(login),
        contentType: "application/json",
        complete: function (xmlHttpRequest, status) {
            var token = xmlHttpRequest.getResponseHeader('Session-Token');
            sessionStorage.setItem('token', token);
        }
    }).done(success).fail(function (x, msg) {
        failure("login failed: " + x.getResponseHeader('Error-Message'));
    });
}

function service_logout(success, failure) {
    $.ajax({
        type: "DELETE",
        url: baseUrl + "token/" + sessionStorage.token
    }).done(function () {
        sessionStorage.setItem('token', null);
        success();
    }).fail(function (x, msg) { failure("logout failed: " + msg); });
}

function list_sessions(success, failure) {
    $.ajax({
        type: "GET",
        url: baseUrl + "sessions",
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function (data) {
        var cleanData = data.map(function (session) {
            return { userId: session.userId, userName: session.user.name, roomId: session.roomId, lastTouch: session.lastTouch }
        })
        success(cleanData)
    }).fail(function (x, msg) {
        failure("list sessions failed: " + x.getResponseHeader('Error-Message'));
    });
}

//[<WebInvoke(Method = "PUT", UriTemplate="users", RequestFormat = WebMessageFormat.Json, 
//            ResponseFormat = WebMessageFormat.Json, BodyStyle = WebMessageBodyStyle.Bare)>]
//[<OperationContract>]
//member this.AddUser(user:User) =
//this.ImplementationWrapper((fun () -> data.AddUser(getToken(),user)),"")

//[<WebInvoke(Method = "DELETE", UriTemplate="users/{userIdNumPart}")>]
//[<OperationContract>]
//member this.RemoveUser(userIdNumPart:string) =
//this.ImplementationWrapper((fun () -> data.RemoveUser(getToken(),"users/"+userIdNumPart)),())

//[<WebGet(UriTemplate="users",ResponseFormat=WebMessageFormat.Json, BodyStyle=WebMessageBodyStyle.Bare)>]
//[<OperationContract>]
//member this.ListUsers() =
//    this.ImplementationWrapper((fun () -> data.ListUsers(getToken())),Array.empty)
                                        
//[<WebInvoke(Method = "PUT", UriTemplate="rooms", RequestFormat = WebMessageFormat.Json, 
//            ResponseFormat = WebMessageFormat.Json, BodyStyle = WebMessageBodyStyle.Bare)>]
//[<OperationContract>]
//member this.AddRoom(room:Room) =
//this.ImplementationWrapper((fun () -> data.AddRoom(getToken(),room)),"")

//[<WebInvoke(Method = "DELETE", UriTemplate="rooms/{roomIdNumPart}")>]
//[<OperationContract>]
//member this.RemoveRoom(roomIdNumPart:string) =
//this.ImplementationWrapper((fun () -> data.RemoveRoom(getToken(),"rooms/"+roomIdNumPart)),())



function list_rooms(success,failure) {
    $.ajax({
        type: "GET",
        url: baseUrl + "rooms",
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function (data) {
        success(data)
    }).fail(function (x, msg) {
        failure("list rooms failed: " + x.getResponseHeader('Error-Message'));
    });
}

function enter_room(roomId,success,failure) {
    var params = { id: roomId };

    $.ajax({
        type: "PUT",
        url: baseUrl + "currentroom",
        data: JSON.stringify(params),
        contentType: "application/json",
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function () {
        success("entered room: " + roomId);
    }).fail(function (x, msg) {
        failure("enter room failed: " + x.getResponseHeader('Error-Message'));
    });
}

function leave_room(success, failure) {
    $.ajax({
        type: "DELETE",
        url: baseUrl + "currentroom",
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function () {
        success("left the room");
    }).fail(function (x, msg) {
        failure("leave room failed: " + x.getResponseHeader('Error-Message'));
    });
}

function get_messages(success, failure) {
    var date = moment().subtract('days', 14).format("YYYY-MM-DDTHH:mm:ss")

    $.ajax({
        type: "GET",
        url: baseUrl + "messages?after=" + date,
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function (data) {
        var cleanData = data.map(function (msg) {
            return { userName: msg.userName, timestamp: msg.timestamp, rawMessage: msg.rawMessage, html: msg.html }
        })
        success(cleanData)
    }).fail(function (x, msg) {
        failure("get messages failed: " + x.getResponseHeader('Error-Message'));
    });
}

function post_message(message, success, failure) {
    var params = { message: message }

    $.ajax({
        type: "PUT",
        url: baseUrl + "messages",
        data: JSON.stringify(params),
        contentType: "application/json",
        headers: { 'Session-Token': sessionStorage.token }
    }).done(function () {
        success();
    }).fail(function (x, msg) {
        failure("post message failed: " + x.getResponseHeader('Error-Message'));
    });
}
